### PR TITLE
Allow retries on Vault MP Config Source initialization

### PR DESCRIPTION
**Pull request by vsevel**
During the initialization of the Vault Config Source, vault (or one specific replica) might be temporarily unavailable.
In that case the rest call through the vertx web client will throw a mutiny timeout exception, which will make the application fail at startup.
Retrying, sometimes just once, may be enough to successfully fetch the secrets from the Vault (or a good replica) after an initial failure, which will allow for the application to continue its initialization, and avoid a crash and a restart.
Ideally the remote service should always be up. But this can't be guaranteed, and since it is on the critical path of application startup, it is fair to add some resiliency.
This PR provides a way to allow some retries during the very first time secrets get fetched by the vault config source.
Once fetched successfully once, the retries will not be used again. Instead we will rely on the internal cache.
Retries are used only in the context of the vault config source. In particular it is not being used by the different programmatic interfaces for the supported engines.
To summarize it is only used by the vault config source, and only on the very first access.

### CODE REVIEW

**Review by geoand — APPROVED** [review]
LGTM

**Comment by vsevel** [comment]
not ready to merge. I am reworking the impl.

**Comment by vsevel** [comment]
hello @geoand . thanks for the initial review. there were a few shortcomings I realized and wanted to address:

- it was doing the retries on the _read secret_ operation, when the very first call is actually the login. so the initial impl would have failed on the login first
- I was retrying only on read timeouts, not on idle timeout (connection gets closed), cannot connect, or connection closed (e.g. reset by peer)

so I changed the design to translate the different error conditions into a common exception `VaultIOException` and handled it directly into the config source.

I also changed the default read timeout to be `5S` instead of `1S`, which was largely insufficient in case of a temporary slow down.

does this sound reasonable? thanks for your time.

**Comment by geoand** [comment]
I'll let @sberyozkin review this as well

**Comment by vsevel** [comment]
hi @sberyozkin how are you? any chance you could take a look? thanks.

**Comment by vsevel** [comment]
hi @sberyozkin do you think you will be able to take a look? or if you do not have time, would you know somebody who may be able to do it? thanks.

**Review by sberyozkin — APPROVED** [review]
